File: lib/core/models/student.dart
class Student {
  final String id;
  final String admissionNumber;
  final String? rollNumber;
  final String name;
  final String dob;
  final String gender;
  final String classId;
  final String? className;
  final String? section;
  final String fatherName;
  final String motherName;
  final String guardianPhone;
  final String address;
  final String admissionDate;
  final String? photoPath;
  final String status;
  final String academicSessionId;
  final double pendingAmount;

  Student({
    required this.id,
    required this.admissionNumber,
    this.rollNumber,
    required this.name,
    required this.dob,
    required this.gender,
    required this.classId,
    this.className,
    this.section,
    required this.fatherName,
    required this.motherName,
    required this.guardianPhone,
    required this.address,
    required this.admissionDate,
    this.photoPath,
    required this.status,
    required this.academicSessionId,
    this.pendingAmount = 0.0,
  });

  factory Student.fromJson(Map<String, dynamic> json) {
    return Student(
      id: json['_id'] ?? json['id'] ?? '',
      admissionNumber: json['admissionNumber'] ?? '',
      rollNumber: json['rollNumber'],
      name: json['name'] ?? '',
      dob: json['dob'] ?? '',
      gender: json['gender'] ?? '',
      classId: json['currentClass'] is Map
          ? json['currentClass']['_id']
          : (json['currentClass'] ?? ''),
      className: json['currentClass'] is Map
          ? json['currentClass']['name']
          : json['className'],
      section: json['section'],
      fatherName: json['fatherName'] ?? '',
      motherName: json['motherName'] ?? '',
      guardianPhone: json['guardianPhone'] ?? json['contact'] ?? '',
      address: json['address'] ?? '',
      admissionDate: json['admissionDate'] ?? '',
      photoPath: json['photoPath'],
      status: json['status'] ?? 'active',
      academicSessionId: json['academicSession'] is Map
          ? json['academicSession']['_id']
          : (json['academicSession'] ?? ''),
      pendingAmount: double.tryParse(json['pendingAmount']?.toString() ?? '0') ?? 0.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'admissionNumber': admissionNumber,
      'rollNumber': rollNumber,
      'name': name,
      'dob': dob,
      'gender': gender,
      'currentClass': classId,
      'section': section,
      'fatherName': fatherName,
      'motherName': motherName,
      'guardianPhone': guardianPhone,
      'address': address,
      'admissionDate': admissionDate,
      'photoPath': photoPath,
      'status': status,
      'academicSession': academicSessionId,
      'pendingAmount': pendingAmount,
    };
  }

  Student copyWith({
    String? id,
    String? admissionNumber,
    String? rollNumber,
    String? name,
    String? dob,
    String? gender,
    String? classId,
    String? className,
    String? section,
    String? fatherName,
    String? motherName,
    String? guardianPhone,
    String? address,
    String? admissionDate,
    String? photoPath,
    String? status,
    String? academicSessionId,
    double? pendingAmount,
  }) {
    return Student(
      id: id ?? this.id,
      admissionNumber: admissionNumber ?? this.admissionNumber,
      rollNumber: rollNumber ?? this.rollNumber,
      name: name ?? this.name,
      dob: dob ?? this.dob,
      gender: gender ?? this.gender,
      classId: classId ?? this.classId,
      className: className ?? this.className,
      section: section ?? this.section,
      fatherName: fatherName ?? this.fatherName,
      motherName: motherName ?? this.motherName,
      guardianPhone: guardianPhone ?? this.guardianPhone,
      address: address ?? this.address,
      admissionDate: admissionDate ?? this.admissionDate,
      photoPath: photoPath ?? this.photoPath,
      status: status ?? this.status,
      academicSessionId: academicSessionId ?? this.academicSessionId,
      pendingAmount: pendingAmount ?? this.pendingAmount,
    );
  }
}

File: Backend/models/Student.js
const mongoose = require('mongoose');

const studentSchema = mongoose.Schema(
  {
    admissionNumber: {
      type: String,
      required: true,
    },
    rollNumber: {
      type: String,
    },
    name: {
      type: String,
      required: true,
    },
    dob: {
      type: String,
      required: true,
    },
    gender: {
      type: String,
      required: true,
    },
    currentClass: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Class',
      required: true,
    },
    section: {
      type: String,
    },
    fatherName: {
      type: String,
      required: true,
    },
    motherName: {
      type: String,
      required: true,
    },
    guardianPhone: {
      type: String,
      required: true,
    },
    address: {
      type: String,
      required: true,
    },
    admissionDate: {
      type: String,
      required: true,
    },
    photoPath: {
      type: String,
    },
    status: {
      type: String,
      enum: ['active', 'inactive', 'graduated'],
      default: 'active',
    },
    academicSession: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'AcademicSession',
      required: true,
    },
  },
  {
    timestamps: true,
  }
);

// 1. Unique index to prevent duplicate roll numbers within a class/section/session
studentSchema.index(
  { academicSession: 1, currentClass: 1, section: 1, rollNumber: 1 },
  { unique: true, partialFilterExpression: { rollNumber: { $type: 'string' } } }
);

// 2. Scoped unique index for admissionNumber within the AcademicSession
studentSchema.index(
  { academicSession: 1, admissionNumber: 1 },
  { unique: true }
);

// Pre-save hook for validation
studentSchema.pre('save', async function () {
  if (this.isModified('currentClass') || this.isModified('section') || this.isModified('academicSession')) {
    const Class = mongoose.model('Class');
    const cls = await Class.findById(this.currentClass);

    if (!cls) {
      throw new Error('Referenced class does not exist');
    }

    // 3. Verify that currentClass actually belongs to the given academicSession
    if (String(cls.academicSession) !== String(this.academicSession)) {
      throw new Error('Class does not belong to the selected academic session');
    }

    // 2. Verify that section is valid for the class if the class has sections
    if (cls.sections && cls.sections.length > 0) {
      if (!this.section) {
        throw new Error(`Section is required for class '${cls.name}'`);
      }
      const validSection = cls.sections.some((s) => s.name === this.section);
      if (!validSection) {
        throw new Error(`Section '${this.section}' does not exist in class '${cls.name}'`);
      }
    } else {
      // If the class has no sections, the section field should be empty
      this.section = undefined;
    }
  }
});

module.exports = mongoose.model('Student', studentSchema);
